lisa/sut_orchestrator/azure/features.py

```diff
@@ -108,10 +121,209 @@ def _execute(self, wait: bool, operator: str, **kwargs: Any) -> None:
+            "api_version", api_version, "str"
+        )
+
+        # Construct headers
+        header_parameters = {}  # type: Dict[str, Any]
+
+        # Fix SerialPortsOperations: Add Content-Type header
+        header_parameters["Content-Type"] = self._serialize.header(
+            "content_type", content_type, "str"
+        )
+        header_parameters["Accept"] = self._serialize.header("accept", accept, "str")
+
+        request = self._client.post(url, query_parameters, header_parameters)
+        pipeline_response = self._client._pipeline.run(request, stream=False, **kwargs)
+        response = pipeline_response.http_response
+
+        if response.status_code not in [200]:
+            map_error(  # type: ignore
+                status_code=response.status_code, response=response, error_map=error_map
+            )
+            raise HttpResponseError(
+                response=response, error_format=ARMErrorFormat
+            )  # type: ignore
+
+        deserialized = self._deserialize("SerialPortConnectResult", pipeline_response)
+
+        if cls:
+            return cls(pipeline_response, deserialized, {})
+
+        return deserialized
+
+    connect.metadata = {  # type: ignore
+        "url": (
+            "/subscriptions/{subscriptionId}/"
+            "resourcegroups/{resourceGroupName}/"
+            "providers/{resourceProviderNamespace}/"
+            "{parentResourceType}/{parentResource}/"
+            "providers/Microsoft.SerialConsole/"
+            "serialPorts/{serialPort}/connect"
+        )
+    }
+
+
 class SerialConsole(AzureFeatureMixin, features.SerialConsole):
+    RESOURCE_PROVIDER_NAMESPACE = "Microsoft.Compute"
+    PARENT_RESOURCE_TYPE = "virtualMachines"
+    DEFAULT_SERIAL_PORT_ID = 0
+
     def _initialize(self, *args: Any, **kwargs: Any) -> None:
         super()._initialize(*args, **kwargs)
         self._initialize_information(self._node)
+        self._initialize_serial_console(id=self.DEFAULT_SERIAL_PORT_ID)
+
+    def write(self, cmd: str, timeout: int = 120) -> None:
+        retry = 0
+        for retry in range(3):
+            # websocket connection is not stable, so we need to retry
+            try:
+                self._write(cmd)
+                return
+            except websockets.ConnectionClosed as e:  # type: ignore
+                # If the connection is closed, we need to reconnect
+                self._log.debug(f"Connection closed: {e}")
+                self._ws = None
+                self._get_connection()
+                last_raised_excption = e
+            except Exception as e:
+                self._log.debug(f"Failed to run command: {e}")
+                last_raised_excption = e
+
+            retry += 1
+
+        raise LisaException(
+            f"Failed to write command to serial console: {last_raised_excption}"
+        )
+
+    def read(self, timeout: int = 120) -> str:
+        retry = 0
+        for retry in range(3):
```

Comment:
It looks the whole method is for retry, the "retry" decorator can be used directly.